contrib/mesos/pkg/minion/task.go

```diff
@@ -0,0 +1,519 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package minion
+
+import (
+	"fmt"
+	"io"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+
+	log "github.com/golang/glog"
+)
+
+const defaultTaskRestartDelay = 5 * time.Second
+
+type taskCompletion struct {
+	name string // name of the task
+	code int    // exit code that the task process completed with
+}
+
+// systemProcess is a useful abstraction for testing
+type systemProcess interface {
+	// Wait works like exec.Cmd.Wait()
+	Wait() error
+
+	// Kill returns the pid of the process that was killed
+	Kill() (int, error)
+}
+
+type cmdProcess struct {
+	delegate *exec.Cmd
+}
+
+func (cp *cmdProcess) Wait() error {
+	return cp.delegate.Wait()
+}
+
+func (cp *cmdProcess) Kill() (int, error) {
+	// TODO(jdef) SIGKILL is pretty harsh, should we implement kill escalation: start with TERM
+	// and follow w/ KILL in x seconds?
+
+	// kill the entire process group, not just the one process
+	pid := cp.delegate.Process.Pid
+	processGroup := 0 - pid
+	rc := syscall.Kill(processGroup, syscall.SIGKILL)
+	return pid, rc
+}
+
+// task is a specification for running a system process; it provides hooks for customizing
+// logging and restart handling as well as provides event channels for communicating process
+// termination and errors related to process management.
+type task struct {
+	name         string                // required: unique name for this task
+	bin          string                // required: path to executable
+	args         []string              // optional: process arguments
+	env          []string              // optional: process environment override
+	createLogger func() io.WriteCloser // factory func that builds a log writer
+	cmd          systemProcess         // process that we started
+	errCh        chan error            // errCh reports errors encountered while trying to execute task processes
```

Comment:
But are there actually multiple errors per launch? It introduces a lot of complexity. I think it's worth a try to combine them. I expect a lot of code to go a way.